bundles/org.openhab.binding.enigma2/src/main/java/org/openhab/binding/enigma2/internal/Enigma2Client.java

```diff
@@ -0,0 +1,289 @@
+    static final String PATH_MESSAGE = "/web/message?type=";
+    static final String PATH_ALL_SERVICES = "/web/getallservices";
+    static final String PATH_ZAP = "/web/zap?sRef=";
+    static final String PATH_CHANNEL = "/web/subservices";
+    static final String PATH_EPG = "/web/epgservice?sRef=";
+    static final int TYPE_INFO = 1;
+    static final int TYPE_WARNING  = 2;
+    static final int TYPE_ERROR = 3;
+    private Map<String,@Nullable String> channels = new LinkedHashMap<>();
+    private String host;
+    private boolean power;
+    private String channel = "";
+    private String title = "";
+    private String description = "";
+    private int volume = 0;
+    private int timeout;
+    private boolean mute;
+    private boolean online;
+    private boolean initialized = false;
+
+    public Enigma2Client(String host, @Nullable String user, @Nullable String password, int timeout) {
+        this.timeout = timeout;
+        if(StringUtils.isNotEmpty(user) && StringUtils.isNotEmpty(password)) {
+            this.host = "http://" + user + ":" + password + "@" + host;
+        } else {
+            this.host = "http://" + host;
+        }
+    }
+
+    public boolean refresh() {
+        refreshPower();
+        refreshAllServices();
+        refreshChannel();
+        refreshEpg();
+        refreshVolume();
+        return online;
+    }
+
+    public void refreshPower() {
+        Optional<Document> document = transmit(PATH_POWER);
+        if(document.isPresent()) {
+            online = true;
+            processPowerResult(document.get());
+        } else {
+            online = false;
+            power = false;
+        }
+        initialized = true;
+    }
+
+    public void refreshAllServices() {
+        if(power || channels.isEmpty()) {
+            transmit(PATH_ALL_SERVICES).ifPresent(this::processAllServicesResult);
+        }
+    }
+
+    public void refreshChannel() {
+        if(power) {
+            transmit(PATH_CHANNEL).ifPresent(this::processChannelResult);
+        }
+    }
+
+    public void refreshVolume() {
+        if(power) {
+            transmit(PATH_VOLUME).ifPresent(this::processVolumeResult);
+        }
+    }
+
+    public void refreshEpg() {
+        if(power) {
+            Optional.ofNullable(channels.get(channel)).ifPresent(name -> {
+                transmit(PATH_EPG + name).ifPresent(this::processEpgResult);
+            });
+        }
+    }
+
+    private Optional<Document> transmit(String path) {
+        try {
+            DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
+            DocumentBuilder builder = factory.newDocumentBuilder();
+            getLogger().debug("Transmitting {}", host + path);
+            String xml = getEnigma2HttpClient().get(host + path);
+            getLogger().debug("Transmitting result is {}", xml);
+            return Optional.ofNullable(builder.parse(new InputSource(new StringReader(xml))));
+        } catch (Exception e) {
```

Comment:
Its in general a bad habit to catch `Exception` directly (besides other things this will also swallow RuntimeExceptions like NPEs for example), please catch the more specific ones.

Reply:
Fixed